flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/ApiRules.java

```diff
@@ -0,0 +1,93 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils.architecture;
+
+import org.apache.flink.annotation.Public;
+import org.apache.flink.annotation.PublicEvolving;
+
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+
+import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideOutsideOfPackage;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+import static org.apache.flink.testutils.architecture.Predicates.areAnnotatedWithAtLeastOneOf;
+import static org.apache.flink.testutils.architecture.Predicates.haveLeafArgumentTypes;
+import static org.apache.flink.testutils.architecture.Predicates.haveLeafReturnTypes;
+
+/** Architecture rules for APIs. */
+public class ApiRules {
+
+    /**
+     * Ensure that public methods marked {@link Public} only use return and argument types which are
+     * also marked {@link Public}.
+     *
+     * <p>Note that this is a stronger requirement than {@link
+     * #PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES} and ensures that {@link
+     * Public} APIs do not use {@link PublicEvolving} APIs.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_API_METHODS_USE_ONLY_PUBLIC_API_TYPES =
+            methods()
+                    .that()
+                    .areDeclaredInClassesThat()
+                    .areAnnotatedWith(Public.class)
+                    .and()
+                    .arePublic()
+                    .should(
+                            haveLeafReturnTypes(
+                                    resideOutsideOfPackage("org.apache.flink..")
+                                            .or(
+                                                    areAnnotatedWithAtLeastOneOf(
+                                                            Public.class, Deprecated.class))))
+                    .andShould(
+                            haveLeafArgumentTypes(
+                                    resideOutsideOfPackage("org.apache.flink..")
+                                            .or(
+                                                    areAnnotatedWithAtLeastOneOf(
+                                                            Public.class, Deprecated.class))));
+
+    /**
+     * Ensure that public methods marked {@link Public} or {@link PublicEvolving} only use return
+     * and argument types which are also marked {@link Public} or {@link PublicEvolving}.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
```

Comment:
We could also think about making these stricter: instead of only checking argument/return types which reside in o.a.f, check any classes which do not reside in java.*. This might require some additional allow-list. It's a question of how much we want to allow potentially relying on classes which are neither from the SDK nor Flink.

Reply:
Not a review, I was just peaking around at how this works. 

In general the only allow list is:
- org.apache.flink.* but _not_ org.apache.flink.shaded (!)
- java.*
- scala.* (only in the scala api's)
- Kryo https://github.com/apache/flink/blob/99c2a415e9eeefafacf70762b6f54070f7911ceb/flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java#L720-L812 

And then connectors will often expose some number of classes from their underlying client. For example, the Kafka serde classes expose ConsumerRecord and ProducerRecord.

Reply:
Thanks, Seth, that sounds like a good list. I think shaded and Scala we would then consider directly in the rule; for Kryo and connector-specific things I would probably allow the violation and freeze them. Being aware and made aware on changes that further expose third-party classes would be valuable, and a connector-specific allow-list would be needed either way anyway.